File: sw/rust/map/src/coord_map_n.rs
use tagma_core::CoordSpaceN;

use crate::coord_gen::CoordKey;
use crate::{CoordMap, CoordMapKey};

// ── Internal helpers ──────────────────────────────────────────────────────

fn vec_to_box(v: Vec<u8>) -> Box<[u8]> {
    v.into_boxed_slice()
}

fn box_to_vec(v: &[u8]) -> Vec<u8> {
    v.to_vec()
}

fn box_to_vec_owned(v: Box<[u8]>) -> Vec<u8> {
    v.into_vec()
}

fn path_to_key_bytes<const N: usize>(path: &tagma_core::CoordPath<N>) -> [u8; N] {
    let mut key = [0u8; N];
    for (i, c) in path.coords().iter().enumerate() {
        key[i] = c.index() as u8;
    }
    key
}

// ── CoordMapN ──────────────────────────────────────────────────────────────

/// A fixed-N-byte-key map backed by [`CoordSpaceN`] — the sparse
/// tree for any depth `N`.
///
/// Keys must be exactly `N` bytes.  Lookup cost is O(N) via tree traversal.
///
/// # Trait implementations
///
/// | Trait | Methods |
/// |-------|---------|
/// | [`CoordMap`] | `insert`, `get`, `remove`, `contains_key` via `&str` |
/// | [`CoordMapKey<N>`] | `insert_by_coordkey`, `get_by_coordkey`, `remove_by_coordkey`, `contains_key_by_coordkey` via `CoordKey<N>` |
///
/// # Example
///
/// ```
/// use tagma_map::{CoordMap, CoordMapKey};
/// use tagma_map::coord_map_n::CoordMapN;
///
/// let mut kv = CoordMapN::<3>::new();
/// kv.insert("foo", b"bar".to_vec());
/// assert_eq!(kv.get("foo"), Some(b"bar".to_vec()));
/// ```
pub struct CoordMapN<const N: usize> {
    space: CoordSpaceN<N, Box<[u8]>>,
    len: usize,
}

impl<const N: usize> CoordMapN<N> {
    /// Creates an empty N-byte-key store.
    pub fn new() -> Self {
        CoordMapN {
            space: CoordSpaceN::new(),
            len: 0,
        }
    }

    /// Returns an iterator over `(key_bytes, value)` pairs in
    /// ascending coordinate order.
    pub fn iter(&self) -> CoordMapNIter<'_, N> {
        CoordMapNIter {
            inner: self.space.iter_tree(),
            _phantom: core::marker::PhantomData,
        }
    }
}

impl<const N: usize> CoordMap for CoordMapN<N> {
    fn len(&self) -> usize {
        self.len
    }

    fn is_empty(&self) -> bool {
        self.len == 0
    }

    fn clear(&mut self) {
        self.space.clear();
        self.len = 0;
    }

    fn insert(&mut self, key: &str, value: Vec<u8>) -> Option<Vec<u8>> {
        // Panics if key.len() != N (via CoordKey::from)
        let ck: CoordKey<N> = key.into();
        self.insert_by_coordkey(&ck, value)
    }

    fn get(&self, key: &str) -> Option<Vec<u8>> {
        if key.len() != N {
            return None;
        }
        let ck: CoordKey<N> = key.into();
        self.get_by_coordkey(&ck)
    }

    fn remove(&mut self, key: &str) -> Option<Vec<u8>> {
        if key.len() != N {
            return None;
        }
        let ck: CoordKey<N> = key.into();
        self.remove_by_coordkey(&ck)
    }
}

impl<const N: usize> CoordMapKey<N> for CoordMapN<N> {
    fn insert_by_coordkey(&mut self, key: &CoordKey<N>, value: Vec<u8>) -> Option<Vec<u8>> {
        let path = key.to_coord_path();
        let prev = self.space.place_path(&path, vec_to_box(value));
        if prev.is_none() {
            self.len += 1;
        }
        prev.map(box_to_vec_owned)
    }

    fn get_by_coordkey(&self, key: &CoordKey<N>) -> Option<Vec<u8>> {
        let path = key.to_coord_path();
        self.space.at_path(&path).map(|v| box_to_vec(v.as_ref()))
    }

    fn remove_by_coordkey(&mut self, key: &CoordKey<N>) -> Option<Vec<u8>> {
        let path = key.to_coord_path();
        let val = self.space.vacate_path(&path).map(box_to_vec_owned);
        if val.is_some() {
            self.len = self.len.saturating_sub(1);
        }
        val
    }
}

impl<const N: usize> Default for CoordMapN<N> {
    fn default() -> Self {
        Self::new()
    }
}

// ── Iterator ──────────────────────────────────────────────────────────────

/// An iterator over [`CoordMapN<N>`] entries, yielding `(key_bytes, value)`.
pub struct CoordMapNIter<'a, const N: usize> {
    inner: tagma_core::coord_space_n::TreeIter<'a, N, Box<[u8]>>,
    _phantom: core::marker::PhantomData<[u8; N]>,
}

impl<'a, const N: usize> Iterator for CoordMapNIter<'a, N> {
    type Item = ([u8; N], &'a [u8]);

    fn next(&mut self) -> Option<Self::Item> {
        let (path, val) = self.inner.next()?;
        let key = path_to_key_bytes(&path);
        Some((key, val.as_ref()))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.inner.size_hint()
    }
}

File: sw/java/tagma-map/src/main/java/org/ssccs/syntagma/map/CoordMapN.java
package org.ssccs.syntagma.map;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

import org.ssccs.syntagma.core.CoordPath;
import org.ssccs.syntagma.core.CoordSpaceN;

/**
 * A fixed-depth byte-key map over the {@link CoordSpaceN} tree, the sparse
 * store for any depth.
 *
 * <p>Keys are exactly {@code depth} UTF-8 bytes long, and lookup cost is
 * {@code O(depth)} tree traversal with no hashing and no collisions. The depth
 * is immutable instance state validated by the constructor, the Java form of
 * the C++ and Rust template parameter {@code N}.
 *
 * <p>The entry count is delegated to {@link CoordSpaceN#size()}, which counts
 * the same placement and removal events as the explicit counter of the
 * references.
 *
 * <p>The class is not final because {@link CoordMap2} fixes the depth at 2,
 * mirroring the C++ alias {@code using CoordMap2 = CoordMapN<2>}.
 *
 * <p>Port of the C++ {@code tagma_map::CoordMapN<N>} in
 * {@code sw/cpp/tagma_map/include/tagma_map/coord_map_n.h}; the underlying
 * behavior mirrors the Rust {@code CoordMapN<N>} in
 * {@code sw/rust/map/src/coord_map_n.rs}.
 */
public class CoordMapN implements CoordMapKey, CoordPathLookup {

    private final CoordSpaceN<byte[]> space;

    /**
     * Creates an empty store of {@code depth} levels.
     *
     * @param depth the exact key byte length, at least 1
     * @throws IllegalArgumentException when {@code depth} is less than 1
     */
    public CoordMapN(int depth) {
        this.space = new CoordSpaceN<>(depth);
    }

    // ------------------------------------------------------------------
    // String-key API
    // ------------------------------------------------------------------

    /**
     * {@inheritDoc}
     *
     * <p>The key must be exactly {@code depth} UTF-8 bytes long; a mismatch
     * throws instead of storing, mirroring the Rust panic and the C++
     * {@code std::invalid_argument}.
     */
    @Override
    public Optional<byte[]> insert(String key, byte[] value) {
        Objects.requireNonNull(key, "key");
        Objects.requireNonNull(value, "value");
        return insertByCoordKey(CoordKey.fromString(key, space.depth()), value);
    }

    /**
     * {@inheritDoc}
     *
     * <p>A key whose UTF-8 byte length is not {@code depth} is absent by
     * definition, mirroring the length guard of the references.
     */
    @Override
    public Optional<byte[]> get(String key) {
        Objects.requireNonNull(key, "key");
        byte[] bytes = key.getBytes(StandardCharsets.UTF_8);
        if (bytes.length != space.depth()) {
            return Optional.empty();
        }
        return getByCoordKey(new CoordKey(bytes));
    }

    /**
     * {@inheritDoc}
     *
     * <p>A key whose UTF-8 byte length is not {@code depth} is absent by
     * definition, mirroring the length guard of the references.
     */
    @Override
    public Optional<byte[]> remove(String key) {
        Objects.requireNonNull(key, "key");
        byte[] bytes = key.getBytes(StandardCharsets.UTF_8);
        if (bytes.length != space.depth()) {
            return Optional.empty();
        }
        return removeByCoordKey(new CoordKey(bytes));
    }

    // ------------------------------------------------------------------
    // CoordKey API
    // ------------------------------------------------------------------

    @Override
    public Optional<byte[]> insertByCoordKey(CoordKey key, byte[] value) {
        Objects.requireNonNull(key, "key");
        Objects.requireNonNull(value, "value");
        return space.placePath(key.toCoordPath(), value.clone());
    }

    @Override
    public Optional<byte[]> getByCoordKey(CoordKey key) {
        Objects.requireNonNull(key, "key");
        return space.atPath(key.toCoordPath()).map(byte[]::clone);
    }

    @Override
    public Optional<byte[]> removeByCoordKey(CoordKey key) {
        Objects.requireNonNull(key, "key");
        return space.vacatePath(key.toCoordPath()).map(byte[]::clone);
    }

    // ------------------------------------------------------------------
    // Path lookup and iteration
    // ------------------------------------------------------------------

    /**
     * {@inheritDoc}
     *
     * <p>A path whose length differs from the depth of the store is absent
     * rather than an error, which is what {@link CoordPathLookup} documents.
     * The generated paths of a {@link CoordCubeMap} query follow the query
     * geometry, so a center that cannot address this store yields no hits
     * instead of an exception raised from inside the query loop.
     * {@link CoordSpaceN} itself keeps the rejection for callers that address
     * it directly.
     */
    @Override
    public Optional<byte[]> getByCoordPath(CoordPath path) {
        Objects.requireNonNull(path, "path");
        if (path.length() != space.depth()) {
            return Optional.empty();
        }
        return space.atPath(path).map(byte[]::clone);
    }

    /**
     * All {@code (key, value)} pairs in depth-first coordinate-ascending order.
     * The references hand out pointers into the store; Java hands out copies,
     * so an iterated value is never shared with the store.
     */
    public List<Map.Entry<CoordKey, byte[]>> iter() {
        List<Map.Entry<CoordPath, byte[]>> entries = space.entries();
        List<Map.Entry<CoordKey, byte[]>> out = new ArrayList<>(entries.size());
        for (Map.Entry<CoordPath, byte[]> entry : entries) {
            out.add(Map.entry(CoordKey.fromCoordPath(entry.getKey()), entry.getValue().clone()));
        }
        return out;
    }

    // ------------------------------------------------------------------
    // Size and clearing
    // ------------------------------------------------------------------

    @Override
    public int len() {
        return space.size();
    }

    @Override
    public void clear() {
        space.clear();
    }
}
